rfd/0057-automatic-aws-server-discovery.md

```diff
@@ -235,6 +235,121 @@ discovery_service:
     teleport.dev/origin: "cloud"
 ```
 
+
+### Agentless installation
+
+In addition to supporting automatic Teleport agent installation, an agentless option
+will also be supported. This mode will update the OpenSSH CA to use the Teleport CA
+without installing the full Teleport Agent.
+
+This will require changes to Teleport in order to support RBAC of OpenSSH based
+nodes and to support showing the nodes in tsh/web ui.
+
+This mode can be enabled by setting `agentless: true` in the matcher. When the
+matcher includes this, a predefined script for agentless installation will be used for
+the endpoint.
+
+Example agentless config:
+
+```yaml
+discovery_service:
+  enabled: "yes"
+  aws:
+  - types: ["ec2"]
+    regions: ["us-west-1"]
+    tags:
+      "teleport": "yes" # aws tags to match
+    install:
+      agentless: true
+      # default to this as a result of agentless: true
+      script_name: "default-agentless-installer"
+      sshd_config: "/etc/ssh/sshd_config" # default path
+    ssm:
+      # default to this as a result of agentless: true
+      document_name: "TeleportAgentlessDiscoveryInstaller"
+```
+
+
+An agentless specific SSM document will be required. The `teleport discovery bootstrap`
+command will need to be updated to create SSM documents appropriate for agentless discovery.
+
+Example SSM document:
+```yaml
+# name: TeleportAgentlessDiscoveryInstaller
+---
+schemaVersion: '2.2'
+description: aws:runShellScript
+parameters:
+  caKey:
+    types: String
+    description: "(Required) The Teleport CA to sshd will trust."
+  hostCert:
+    types: String
+    description: "(Required) The Teleport host cert to use."
+  sshdConfigPath:
+    types: String
+    description: "(Required) The path to the sshd config file."
+mainSteps:
+- action: aws:downloadContent
+  name: downloadContent
+  inputs:
+    sourceType: "HTTP"
+    destinationPath: "/tmp/installTeleport.sh"
+    sourceInfo:
+      url: "https://teleportcluster.xyz/webapi/scripts/default-agentless-installer"
+- action: aws:runShellScript
+  name: runShellScript
+  inputs:
+    timeoutSeconds: '300'
+    runCommand:
+      - export CA_KEY='{{ caKey }}'
+      - export HOST_CERT='{{ hostCert }}'
+      - export SSHD_CONFIG='{{ sshdConfigPath }}'
+      - /bin/sh /tmp/installTeleport.sh
+```
+
+Agentless mode will serve a different install script resource named
+`default-agentless-installer`. Which will be used to update and restart the sshd
+configuration.
+
+Possible agentless installer script:
+
+```bash
+(
+  flock -n 9 || exit 1
+
+  if grep -q 'TrustedUserCAKeys /etc/ssh/teleport_user_ca.pub' "$SSHD_CONFIG"; then
+	exit 0
+  fi
+
+  echo "$CA_KEY" > /etc/ssh/teleport_user_ca.pub
+  echo 'TrustedUserCAKeys /etc/ssh/teleport_user_ca.pub' >> $SSHD_CONFIG
+
+  echo 'HostKey /etc/ssh/teleport' >> $SSHD_CONFIG
+  echo 'HostCertificate /etc/ssh/teleport-cert.pub' >> $SSHD_CONFIG
+
+  systemctl restart sshd
```

Comment:
Should we do `sshd -t` first to test the config and make sure we don't brick the node?